api/src/views/user.rs

```diff
@@ -35,42 +35,49 @@ impl<'a> View for InitializeUserView<'a> {
+                            new_user
+                                .insert()
+                                .returning(users::all_columns)
+                                .get_result(conn);
 
-                    // Check if the username already exists
-                    let created_user = DbErrorConverter {
-                        unique_violation_to_exists: true,
-                        ..Default::default()
-                    }
-                    .convert(create_user_result)?;
+                        // Check if the username already exists
+                        let created_user = DbErrorConverter {
+                            unique_violation_to_exists: true,
+                            ..Default::default()
+                        }
+                        .convert(create_user_result)?;
 
-                    // We should update exactly 1 row. If not, then either
-                    // the referenced user_provider row is already linked to
-                    // a user, or it doesn't exist. In either case, just
-                    // return a NotFound error.
-                    let updated_rows = diesel::update(
-                        user_providers::table
-                            .find(user_provider_id)
-                            .filter(user_providers::columns::user_id.is_null()),
-                    )
-                    .set(
-                        user_providers::columns::user_id
-                            .eq(Some(created_user.id)),
-                    )
-                    .execute(conn)?;
+                        // We should update exactly 1 row. If not, then either
+                        // the referenced user_provider row is already linked to
+                        // a user, or it doesn't exist. In either case, just
+                        // return a NotFound error.
+                        let updated_rows = diesel::update(
+                            user_providers::table
+                                .find(user_provider_id)
+                                .filter(
+                                    user_providers::columns::user_id.is_null(),
+                                ),
+                        )
+                        .set(
+                            user_providers::columns::user_id
+                                .eq(Some(created_user.id)),
+                        )
+                        .execute(conn)?;
 
-                    if updated_rows == 0 {
-                        Err(ClientError::NotFound { source: None }.into())
-                    } else {
-                        Ok(created_user)
-                    }
-                })
+                        if updated_rows == 0 {
+                            Err(ClientError::NotFound { source: None }.into())
+                        } else {
+                            Ok(created_user)
+                        }
+                    })
```

Comment:
Basically all I did here was convert the default transaction (which is isolation level `READ COMMITTED`) to the next highest isolation level, `READ REPEATABLE`. We do an insert-then-update here, like so:

1. Insert new user
2. Update one row user_providers to point to that new user

So previously there was a potential race condition if this ran twice simultaneously: Both handlers insert, then the first updates the row to point to its user, then the 2nd does the same update. So you end up with a dangling `users` row from the first request.

With `READ REPEATABLE`, the 2nd request that comes through will update the user_providers row, but when it tries to commit, it will fail because the 1st request modified that row.